### ISSUE TITLE
OrleansClient's Initialize method hangs in IISExpress

### ISSUE THREAD

**Issue by jenyayel** [dotnet/orleans#671]
I spent a few hours on it, so I thought it worth sharing. 

I wasn't managed to figure out why, but it only happens in IISExpress (full IIS, console or self-hosted OWIN works prefectly) the constructor of PerformanceCounter which used in [RuntimeStatisticsGroup] (https://github.com/dotnet/orleans/blob/master/src/Orleans/Statistics/RuntimeStatisticsGroup.cs) will hang, preventing initalization of OrleansClient. Didn't find any relevant trace info or event viewer's logs. Looking at PerformanceCounter I only found a usage of Monitor during its construction, so probably some dead lock there.

Anyway, it would be nice to have an ability to disable statistics in dev environment.

**Comment by gabikliot** [comment]
Interesting. We never saw PerformanceCounter hang up.
One way to handle that would be to run `RuntimeStatisticsGroup.InitCpuMemoryCounters` inside `Task.Run` (on thread pool thread) and [wait on it with timeout](https://github.com/dotnet/orleans/blob/master/src/Orleans/Async/TaskExtensions.cs#L89), and if that times out and throw, re-throw and prevent silo from starting.

**Comment by sergeybykov** [comment]
Have you tried running "lodctr /r" to see if maybe the perf counters registry is corrupted?

**Comment by jenyayel** [comment]
@gabikliot, I found out that [people suffer] (https://www.google.de/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=PerformanceCounter+iisexpress) from this problem in different places. But, how would it improve things to initialize perf counters in separate thread and still throw on timeout (which obviously will happen)? Also, I'm not sure that it is a good idea to throw on a thread that not associated with request. 

I noticed that you have [plans to port to coreclr](#368). Wouldn't be a good idea to abstract statistics from implementation and inject `RuntimeStatisticsGroup`?

@sergeybykov, first thing I tried - didn't help. Thanks.

**Comment by gabikliot** [comment]
@jenyayel it will help in way that the running main thread will not block indefinitely forever, but will timeout and throw. So at least your main process will get back to the caller and you will know what the problem was. Currently, based on what you said, it just silently deadlocks, without letting you even know where and why.
So I am basically suggesting deadlock detection technique, and not deadlock prevention or deadlock recovery. Still better than being stuck in a deadlock forever, right?

As for injecting ` RuntimeStatisticsGroup ` for coreclr, yes we do. We did not think yet through all the details of how exactly it will look like. Would you like to suggest a design for that? How exactly and where ` RuntimeStatisticsGroup ` will be injected? Would it cover all usages of PerformanceCounters? 

If we agree on the design, we can suggest it as open issue on the coreclr port list (#368), and maybe someone will volunteer to do it.

**Comment by jenyayel** [comment]
@gabikliot, you are right, but still - having this issue occurring only in IISExpress and there is no way to fix the core of the issue, preventing from Client (not Silo!) starting will not help to developer. You meant something like this, right?
```
internal RuntimeStatisticsGroup()
{
    try
    {
        Task.Run(() =>
        {
            InitCpuMemoryCounters();
        }).WaitWithThrow(INITIALIZATION_TIMEOUT);  
    }
    catch (TimeoutException)
    {
        logger.Warn(ErrorCode.PerfCounterConnectError,
            "Timeout occurred during initialization of CPU & Memory perf counters");
    }          
}
```

But then again, the timeout period to wait just to start the website... I switched to self hosted OWIN, but I think it worth fixing for others. I'm willing to contribute, tell me what you prefer.

P.S. For coreclr design I will need some time to think.

**Comment by gabikliot** [comment]
Yes, something like that. Its just a workaround, so the silo does not get stuck forever in a deadlock. Until a proper fix for PerformanceCounters is found.

**Comment by gabikliot** [comment]
Closed via #673.
